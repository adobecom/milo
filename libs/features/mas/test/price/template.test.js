import { expect } from '@esm-bundle/chai';
import {
    createPriceTemplate,
    createPromoPriceTemplate,
    createPriceWithAnnualTemplate,
    createPromoPriceWithAnnualTemplate,
} from '../../src/price/template.js';

const context = {
    country: 'US',
    language: 'en',
};
const value = {
    formatString: '#0',
    price: 100,
};

const valueAbm = {
    formatString: '#0',
    price: 100,
    commitment: 'YEAR',
    term: 'MONTHLY',
};

const valueDiscount = {
    formatString: '#0',
    price: 80,
    priceWithoutDiscount: 100,
};

const valueDiscountAbm = {
    formatString: '#0',
    price: 80,
    priceWithoutDiscount: 100,
    commitment: 'YEAR',
    term: 'MONTHLY',
};

const valueNotApplicableDiscount = {
    formatString: '#0',
    price: 100,
    priceWithoutDiscount: 100,
};

const root = document.createElement('div');
document.body.append(root);

/* To update index.test.js.snap, inspect body > div, and run copy($0.outherHTML) in the js console, and paste it */
const snapshots = await fetch('test/price/__snapshots__/template.test.js.snap')
    .then((response) => response.text())
    .then(
        (text) =>
            new DOMParser().parseFromString(text, 'text/html').body
                .firstElementChild,
    );

function renderAndComparePrice(id, html) {
    const el = document.createElement('div', { id });
    el.setAttribute('id', id);
    el.innerHTML = html;
    root.append(el);
    const snapshotEl = snapshots.querySelector(`#${id}`);
    expect(el.innerHTML).to.equal(snapshotEl?.innerHTML);
}

describe('function "createPriceTemplate"', () => {
    describe('argument "attributes"', () => {
        it('renders custom attributes having allowed types', function () {
            const template = createPriceTemplate();
            renderAndComparePrice(
                'createPriceTemplate1',
                template(context, value, {
                    string: 's',
                    json: '{ "foo" : "bar" }',
                    number: 1,
                    truthy: true,
                    falsy: false,
                    null: null,
                    undefined: undefined,
                    array: [1, 2, 3],
                    object: { foo: 'bar' },
                }),
            );
        });

        it('does not throw if attributes are missing', () => {
            const template = createPriceTemplate();
            expect(() => template(context, value)).to.not.throw();
        });
    });
});

describe('function "createPromoPriceTemplate"', () => {
    it('displays both new & old prices', function () {
        const template = createPromoPriceTemplate();
        renderAndComparePrice(
            'createPromoPriceTemplate1',
            template(context, valueDiscount, {
                string: 's',
                json: '{ "foo" : "bar" }',
                number: 1,
                truthy: true,
                falsy: false,
                null: null,
                undefined: undefined,
                array: [1, 2, 3],
                object: { foo: 'bar' },
            }),
        );
    });
    it('displays only new price in case old is the same than new', function () {
        const template = createPromoPriceTemplate();
        renderAndComparePrice(
            'createPromoPriceTemplate2',
            template(context, valueNotApplicableDiscount, {
                string: 's',
                json: '{ "foo" : "bar" }',
                number: 1,
                truthy: true,
                falsy: false,
                null: null,
                undefined: undefined,
                array: [1, 2, 3],
                object: { foo: 'bar' },
            }),
        );
    });
});

describe('function "createPriceWithAnnualTemplate"', function () {
    it('displays ABM with annual price', () => {
        const template = createPriceWithAnnualTemplate();
        renderAndComparePrice(
            'createPriceWithAnnualTemplate1',
            template(context, valueAbm, {}),
        );
    });
});

describe('function "createPromoPriceWithAnnualTemplate"', function () {
    it('displays ABM with annual price and old price', () => {
        const template = createPromoPriceWithAnnualTemplate();
        renderAndComparePrice(
            'createPromoPriceWithAnnualTemplate1',
            template(context, valueDiscountAbm, {}),
        );
    });
});
