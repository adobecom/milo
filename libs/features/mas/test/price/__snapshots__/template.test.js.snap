<div><div id="createPriceTemplate1"><span class="price" string="s" json="{ &quot;foo&quot; : &quot;bar&quot; }" number="1" truthy="" aria-label="100"><span class="price-integer">100</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence disabled"></span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span></div><div id="createPromoPriceTemplate1"><span class="price" string="s" json="{ &quot;foo&quot; : &quot;bar&quot; }" number="1" truthy="" aria-label="80"><span class="price-integer">80</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence disabled"></span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span>&nbsp;<span class="price price-strikethrough" string="s" json="{ &quot;foo&quot; : &quot;bar&quot; }" number="1" truthy="" aria-label="Regularly at 100"><span class="price-integer">100</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence disabled"></span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span></div><div id="createPromoPriceTemplate2"><span class="price" string="s" json="{ &quot;foo&quot; : &quot;bar&quot; }" number="1" truthy="" aria-label="100"><span class="price-integer">100</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence disabled"></span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span></div><div id="createPriceWithAnnualTemplate1"><span class="price" aria-label="100 per month"><span class="price-integer">100</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence">/mo</span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span><span class="price-annual-prefix">&nbsp;(</span><span class="price price-annual" aria-label="1200 per year"><span class="price-integer">1200</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence">/yr</span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span><span class="price-annual-suffix">)</span></div><div id="createPromoPriceWithAnnualTemplate1"><span class="price price-strikethrough" aria-label="Regularly at 100 per month"><span class="price-integer">100</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence">/mo</span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span>&nbsp;<span class="price" aria-label="80 per month"><span class="price-integer">80</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence">/mo</span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span><span class="price-annual-prefix">&nbsp;(</span><span class="price price-annual" aria-label="960 per year"><span class="price-integer">960</span><span class="price-decimals-delimiter disabled"></span><span class="price-decimals disabled"></span><span class="price-currency-space disabled"></span><span class="price-currency-symbol disabled"></span><span class="price-recurrence">/yr</span><span class="price-unit-type disabled"></span><span class="price-tax-inclusivity disabled"></span></span><span class="price-annual-suffix">)</span></div></div>
